File: server/git.ts
/**
 * Real git operations for project workspaces.
 * Uses the system git binary via child_process.
 */

import { exec, spawn } from "child_process";
import { promisify } from "util";
import path from "path";
import fs from "fs";

const execAsync = promisify(exec);

export interface GitFileStatus {
  path: string;
  index: string;  // staged status (' ', 'M', 'A', 'D', 'R', '?', etc.)
  worktree: string; // unstaged status
}

export interface GitCommit {
  oid: string;
  shortOid: string;
  message: string;
  author: string;
  email: string;
  timestamp: number;
}

let gitInstalledCache: boolean | null = null;
let gitVersionCache: string | null = null;

/** Check if git is installed on the system */
export async function isGitInstalled(): Promise<{ installed: boolean; version: string | null }> {
  if (gitInstalledCache !== null) {
    return { installed: gitInstalledCache, version: gitVersionCache };
  }
  try {
    const { stdout } = await execAsync("git --version", { timeout: 5000 });
    gitInstalledCache = true;
    gitVersionCache = stdout.trim();
    return { installed: true, version: gitVersionCache };
  } catch {
    gitInstalledCache = false;
    gitVersionCache = null;
    return { installed: false, version: null };
  }
}

/** Reset cache (call after install attempt) */
export function resetGitCache() {
  gitInstalledCache = null;
  gitVersionCache = null;
}

/** Try to install git on the host system */
export async function installGit(): Promise<{ success: boolean; message: string }> {
  if (process.platform === "win32") {
    // Try winget first (Windows 10/11)
    try {
      await execAsync("winget --version", { timeout: 3000 });
      return new Promise((resolve) => {
        const proc = spawn("winget", ["install", "--id", "Git.Git", "-e", "--source", "winget", "--accept-source-agreements", "--accept-package-agreements"], {
          stdio: "pipe",
          shell: true,
        });
        let output = "";
        proc.stdout?.on("data", (d) => { output += d.toString(); });
        proc.stderr?.on("data", (d) => { output += d.toString(); });
        proc.on("exit", (code) => {
          resetGitCache();
          if (code === 0) {
            resolve({ success: true, message: "Git installed successfully via winget. Please restart the server to use it." });
          } else {
            resolve({ success: false, message: `winget install failed (code ${code}): ${output.slice(-500)}` });
          }
        });
      });
    } catch {
      return {
        success: false,
        message: "Could not auto-install git. Please install manually from https://git-scm.com/download/win",
      };
    }
  } else if (process.platform === "darwin") {
    return {
      success: false,
      message: "On macOS, install git via: brew install git, or xcode-select --install",
    };
  } else {
    return {
      success: false,
      message: "On Linux, install git via your package manager: sudo apt install git, sudo dnf install git, etc.",
    };
  }
}

/** Run a git command in a workspace directory */
async function runGit(cwd: string, args: string[], opts: { timeout?: number } = {}): Promise<{ stdout: string; stderr: string; code: number }> {
  return new Promise((resolve) => {
    const proc = spawn("git", args, { cwd, shell: false });
    let stdout = "";
    let stderr = "";
    proc.stdout.on("data", (d) => { stdout += d.toString(); });
    proc.stderr.on("data", (d) => { stderr += d.toString(); });
    const timer = setTimeout(() => {
      try { proc.kill(); } catch {}
    }, opts.timeout ?? 30000);
    proc.on("exit", (code) => {
      clearTimeout(timer);
      resolve({ stdout, stderr, code: code ?? -1 });
    });
    proc.on("error", () => {
      clearTimeout(timer);
      resolve({ stdout, stderr: "git binary not found", code: -1 });
    });
  });
}

/** Check if a directory is the ROOT of a git repository (not just inside one) */
export async function isGitRepo(workDir: string): Promise<boolean> {
  if (!fs.existsSync(workDir)) return false;
  // Check that .git exists directly in this workspace dir (not a parent)
  const dotGit = path.join(workDir, ".git");
  if (!fs.existsSync(dotGit)) return false;
  // Verify it's a valid git dir by running rev-parse
  const { code } = await runGit(workDir, ["rev-parse", "--is-inside-work-tree"]);
  return code === 0;
}

/** Initialize a git repo in a workspace */
export async function gitInit(workDir: string): Promise<{ success: boolean; message: string }> {
  if (!fs.existsSync(workDir)) return { success: false, message: "Workspace not found" };

  const init = await runGit(workDir, ["init", "-b", "main"]);
  if (init.code !== 0) {
    // Older git might not support -b
    const fallback = await runGit(workDir, ["init"]);
    if (fallback.code !== 0) return { success: false, message: fallback.stderr || "git init failed" };
  }

  // Set default user if missing (avoids commit errors)
  const { stdout: name } = await runGit(workDir, ["config", "user.name"]);
  if (!name.trim()) {
    await runGit(workDir, ["config", "user.name", "PiPilot User"]);
  }
  const { stdout: email } = await runGit(workDir, ["config", "user.email"]);
  if (!email.trim()) {
    await runGit(workDir, ["config", "user.email", "user@pipilot.dev"]);
  }

  return { success: true, message: "Repository initialized" };
}

/** Get git status as a list of files */
export async function gitStatus(workDir: string): Promise<GitFileStatus[]> {
  const { stdout, code } = await runGit(workDir, ["status", "--porcelain=v1", "-uall"]);
  if (code !== 0) return [];

  const files: GitFileStatus[] = [];
  const lines = stdout.split("\n").filter(Boolean);
  for (const line of lines) {
    if (line.length < 3) continue;
    const index = line[0];
    const worktree = line[1];
    let filePath = line.slice(3);
    // Handle renamed files: "old -> new"
    if (filePath.includes(" -> ")) {
      filePath = filePath.split(" -> ")[1];
    }
    // Strip surrounding quotes if any
    if (filePath.startsWith('"') && filePath.endsWith('"')) {
      filePath = filePath.slice(1, -1);
    }
    files.push({ path: filePath, index, worktree });
  }
  return files;
}

/** Get current branch name */
export async function gitCurrentBranch(workDir: string): Promise<string> {
  const { stdout, code } = await runGit(workDir, ["branch", "--show-current"]);
  if (code !== 0) return "main";
  return stdout.trim() || "HEAD";
}

/** List all branches */
export async function gitBranches(workDir: string): Promise<string[]> {
  const { stdout, code } = await runGit(workDir, ["branch", "--list", "--format=%(refname:short)"]);
  if (code !== 0) return [];
  return stdout.split("\n").map(s => s.trim()).filter(Boolean);
}

/** Get commit log */
export async function gitLog(workDir: string, limit = 50): Promise<GitCommit[]> {
  const fmt = "%H%x1f%h%x1f%s%x1f%an%x1f%ae%x1f%at";
  const { stdout, code } = await runGit(workDir, ["log", `--format=${fmt}`, `-n${limit}`]);
  if (code !== 0) return [];

  const commits: GitCommit[] = [];
  for (const line of stdout.split("\n").filter(Boolean)) {
    const [oid, shortOid, message, author, email, ts] = line.split("\x1f");
    commits.push({
      oid, shortOid, message, author, email,
      timestamp: parseInt(ts) || 0,
    });
  }
  return commits;
}

/** Get diff for a file (staged or unstaged) */
export async function gitDiff(workDir: string, filePath: string, staged: boolean): Promise<string> {
  const args = ["diff"];
  if (staged) args.push("--staged");
  args.push("--", filePath);
  const { stdout } = await runGit(workDir, args);
  return stdout;
}

/** Get the original (HEAD) content of a file */
export async function gitShowFile(workDir: string, filePath: string): Promise<string> {
  const { stdout, code } = await runGit(workDir, ["show", `HEAD:${filePath}`]);
  if (code !== 0) return "";
  return stdout;
}

/** Stage files (git add) */
export async function gitAdd(workDir: string, files: string[]): Promise<{ success: boolean; message: string }> {
  if (files.length === 0) return { success: true, message: "No files to stage" };
  const { stderr, code } = await runGit(workDir, ["add", "--", ...files]);
  return { success: code === 0, message: stderr || "Staged" };
}

/** Stage all changes */
export async function gitAddAll(workDir: string): Promise<{ success: boolean; message: string }> {
  const { stderr, code } = await runGit(workDir, ["add", "-A"]);
  return { success: code === 0, message: stderr || "All changes staged" };
}

/** Unstage a file (git reset HEAD <file>) */
export async function gitUnstage(workDir: string, files: string[]): Promise<{ success: boolean; message: string }> {
  if (files.length === 0) return { success: true, message: "" };
  // For initial repo (no HEAD yet), use rm --cached
  const { code: hasHead } = await runGit(workDir, ["rev-parse", "HEAD"]);
  let result;
  if (hasHead === 0) {
    result = await runGit(workDir, ["reset", "HEAD", "--", ...files]);
  } else {
    result = await runGit(workDir, ["rm", "--cached", "--", ...files]);
  }
  return { success: result.code === 0, message: result.stderr || "Unstaged" };
}

/** Commit staged changes */
export async function gitCommit(workDir: string, message: string): Promise<{ success: boolean; message: string }> {
  if (!message.trim()) return { success: false, message: "Commit message required" };
  const { stdout, stderr, code } = await runGit(workDir, ["commit", "-m", message]);
  if (code !== 0) {
    return { success: false, message: stderr || stdout || "Commit failed" };
  }
  return { success: true, message: stdout };
}

/** Push to remote */
export async function gitPush(workDir: string, remote = "origin", branch?: string): Promise<{ success: boolean; message: string }> {
  const args = ["push", remote];
  if (branch) args.push(branch);
  const { stdout, stderr, code } = await runGit(workDir, args, { timeout: 60000 });
  return { success: code === 0, message: stdout + stderr };
}

/** Pull from remote */
export async function gitPull(workDir: string, remote = "origin", branch?: string): Promise<{ success: boolean; message: string }> {
  const args = ["pull", remote];
  if (branch) args.push(branch);
  const { stdout, stderr, code } = await runGit(workDir, args, { timeout: 60000 });
  return { success: code === 0, message: stdout + stderr };
}

/** Create a new branch */
export async function gitCreateBranch(workDir: string, name: string): Promise<{ success: boolean; message: string }> {
  const { stderr, code } = await runGit(workDir, ["branch", name]);
  return { success: code === 0, message: stderr || `Branch ${name} created` };
}

/** Checkout a branch */
export async function gitCheckout(workDir: string, branch: string): Promise<{ success: boolean; message: string }> {
  const { stdout, stderr, code } = await runGit(workDir, ["checkout", branch]);
  return { success: code === 0, message: (stdout + stderr).trim() };
}

/** Discard changes to a file (checkout from HEAD) */
export async function gitDiscard(workDir: string, files: string[]): Promise<{ success: boolean; message: string }> {
  if (files.length === 0) return { success: true, message: "" };
  const { stderr, code } = await runGit(workDir, ["checkout", "HEAD", "--", ...files]);
  return { success: code === 0, message: stderr || "Discarded" };
}

/** Get list of remotes */
export async function gitRemotes(workDir: string): Promise<{ name: string; url: string }[]> {
  const { stdout, code } = await runGit(workDir, ["remote", "-v"]);
  if (code !== 0) return [];
  const remotes = new Map<string, string>();
  for (const line of stdout.split("\n").filter(Boolean)) {
    const [name, urlAndType] = line.split(/\s+/);
    if (name && urlAndType && !remotes.has(name)) {
      remotes.set(name, urlAndType);
    }
  }
  return Array.from(remotes.entries()).map(([name, url]) => ({ name, url }));
}

File: src-tauri/src/git.rs
//! Git operations via CLI — Tauri IPC commands.
//!
//! Each command shells out to `git` via `std::process::Command`,
//! keeping the implementation simple and dependency-free.

use serde::{Deserialize, Serialize};
use std::process::Command;

// ── Result types ──

#[derive(Serialize, Deserialize, Clone)]
pub struct GitCheckResult {
    pub installed: bool,
    pub version: Option<String>,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct GitStatus {
    pub is_repo: bool,
    pub branch: String,
    pub staged: Vec<String>,
    pub modified: Vec<String>,
    pub untracked: Vec<String>,
    pub ahead: u32,
    pub behind: u32,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct GitLogEntry {
    pub sha: String,
    pub message: String,
    pub author: String,
    pub date: String,
}

// ── Helpers ──

/// Run a git command in `cwd`, returning (stdout, stderr).
fn run_git(cwd: &str, args: &[&str]) -> Result<String, String> {
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .map_err(|e| format!("Failed to run git: {}", e))?;

    if output.status.success() {
        Ok(String::from_utf8_lossy(&output.stdout).to_string())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        Err(format!("git error: {}", stderr.trim()))
    }
}

// ── Tauri commands ──

#[tauri::command]
pub fn git_check() -> Result<GitCheckResult, String> {
    let output = Command::new("git")
        .arg("--version")
        .output();

    match output {
        Ok(out) if out.status.success() => {
            let version = String::from_utf8_lossy(&out.stdout).trim().to_string();
            Ok(GitCheckResult {
                installed: true,
                version: Some(version),
            })
        }
        _ => Ok(GitCheckResult {
            installed: false,
            version: None,
        }),
    }
}

#[tauri::command]
pub fn git_status(cwd: String) -> Result<GitStatus, String> {
    // Check if it's a repo
    let is_repo = Command::new("git")
        .args(["rev-parse", "--is-inside-work-tree"])
        .current_dir(&cwd)
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false);

    if !is_repo {
        return Ok(GitStatus {
            is_repo: false,
            branch: String::new(),
            staged: vec![],
            modified: vec![],
            untracked: vec![],
            ahead: 0,
            behind: 0,
        });
    }

    // Branch name
    let branch = run_git(&cwd, &["rev-parse", "--abbrev-ref", "HEAD"])
        .unwrap_or_default()
        .trim()
        .to_string();

    // Porcelain status for file lists
    let status_output = run_git(&cwd, &["status", "--porcelain=v1"]).unwrap_or_default();
    let mut staged = Vec::new();
    let mut modified = Vec::new();
    let mut untracked = Vec::new();

    for line in status_output.lines() {
        if line.len() < 3 {
            continue;
        }
        let x = line.chars().nth(0).unwrap_or(' ');
        let y = line.chars().nth(1).unwrap_or(' ');
        let file = line[3..].to_string();

        if x == '?' {
            untracked.push(file);
        } else {
            if x != ' ' && x != '?' {
                staged.push(file.clone());
            }
            if y != ' ' && y != '?' {
                modified.push(file);
            }
        }
    }

    // Ahead/behind
    let mut ahead: u32 = 0;
    let mut behind: u32 = 0;
    if let Ok(ab) = run_git(&cwd, &["rev-list", "--left-right", "--count", "HEAD...@{upstream}"]) {
        let parts: Vec<&str> = ab.trim().split('\t').collect();
        if parts.len() == 2 {
            ahead = parts[0].parse().unwrap_or(0);
            behind = parts[1].parse().unwrap_or(0);
        }
    }

    Ok(GitStatus {
        is_repo: true,
        branch,
        staged,
        modified,
        untracked,
        ahead,
        behind,
    })
}

#[tauri::command]
pub fn git_init(cwd: String) -> Result<(), String> {
    run_git(&cwd, &["init"])?;
    Ok(())
}

#[tauri::command]
pub fn git_add(cwd: String, paths: Vec<String>) -> Result<(), String> {
    let path_refs: Vec<&str> = paths.iter().map(|s| s.as_str()).collect();
    let mut args = vec!["add"];
    args.extend(path_refs);
    run_git(&cwd, &args)?;
    Ok(())
}

#[tauri::command]
pub fn git_commit(cwd: String, message: String) -> Result<String, String> {
    run_git(&cwd, &["commit", "-m", &message])?;
    // Return the SHA of the new commit
    let sha = run_git(&cwd, &["rev-parse", "HEAD"])?
        .trim()
        .to_string();
    Ok(sha)
}

#[tauri::command]
pub fn git_log(cwd: String, count: u32) -> Result<Vec<GitLogEntry>, String> {
    let count_str = format!("-{}", count);
    // Use a delimiter unlikely to appear in commit messages
    let format = "--format=%H\x1f%s\x1f%an\x1f%ai";
    let output = run_git(&cwd, &["log", &count_str, format])?;

    let mut entries = Vec::new();
    for line in output.lines() {
        let parts: Vec<&str> = line.split('\x1f').collect();
        if parts.len() >= 4 {
            entries.push(GitLogEntry {
                sha: parts[0].to_string(),
                message: parts[1].to_string(),
                author: parts[2].to_string(),
                date: parts[3].to_string(),
            });
        }
    }
    Ok(entries)
}

#[tauri::command]
pub fn git_diff(cwd: String, path: Option<String>) -> Result<String, String> {
    let mut args = vec!["diff"];
    if let Some(ref p) = path {
        args.push("--");
        args.push(p.as_str());
    }
    run_git(&cwd, &args)
}

#[tauri::command]
pub fn git_push(cwd: String, remote: String, branch: String) -> Result<(), String> {
    run_git(&cwd, &["push", &remote, &branch])?;
    Ok(())
}

#[tauri::command]
pub fn git_pull(cwd: String, remote: String, branch: String) -> Result<(), String> {
    run_git(&cwd, &["pull", &remote, &branch])?;
    Ok(())
}

#[tauri::command]
pub fn git_branch_list(cwd: String) -> Result<Vec<String>, String> {
    let output = run_git(&cwd, &["branch", "--format=%(refname:short)"])?;
    let branches: Vec<String> = output
        .lines()
        .map(|l| l.trim().to_string())
        .filter(|l| !l.is_empty())
        .collect();
    Ok(branches)
}

#[tauri::command]
pub fn git_branch_create(cwd: String, name: String) -> Result<(), String> {
    run_git(&cwd, &["branch", &name])?;
    Ok(())
}

#[tauri::command]
pub fn git_checkout(cwd: String, ref_name: String) -> Result<(), String> {
    run_git(&cwd, &["checkout", &ref_name])?;
    Ok(())
}
